README.txt
-------------------------------------------------------------------------------
This explains the use of encrypted EBS-backed root filesystems in Amazon EC2.
Some of the code was originally written to allow use of EBS-backed instances
before Amazon released that feature. 1.5 years later, the original method is
now obsolete. However, any paranoid people out there may still be interested
in a working encryption layer. Encrypting the whole filesystem is conceptually
simpler than having to deal with separate encrypted volumes, and you can stop
worrying about potential secrets leaking through configuration or log files.

Of course, the hardware is still physically beyond your control, so it is up
to you to decide whether encryption adds any real value. On the other hand,
these steps should make it easy to construct your own instance, and you never
know if some sensitive backup tape will end up in the wrong hands.

                                     - Henrik Gulbrandsen <henrik@gulbra.net>
                                               2011-04-18

-------------------------------------------------------------------------------
I have updated the text below to match current reality. It is still valid, but
you should know that it only shows what you could do three years ago, which
was to create an encrypted classic EC2 instance with a clean Ubuntu system.

Today, you can also create Debian systems, clone an existing instance, start
VPC instances with multiple network interfaces and groups, use an IP address
as the "domain" if you really want to, add extra call parameters for advanced
EC2 features, and hook in your own script to handle a favorite Linux distro.
At the very least, you should probably look up the --name and --size options.

You can also install Encroot on instances with "./configure && make install",
which may be a good idea if you intend to use it regularly.

For more detailed info, please read the encroot(1) and awsapi(1) man pages.

                                     - Henrik Gulbrandsen <henrik@gulbra.net>
                                               2014-06-05

-------------------------------------------------------------------------------

This was tested with ami-41e0b93b (64-bit EBS us-east-1 - Ubuntu 16.04 LTS)
See https://help.ubuntu.com/community/UEC/Images for info on Ubuntu images.
Latest Xenial Xerus: https://uec-images.ubuntu.com/releases/xenial/release/

 A. In this setup, the decryption password is given via HTTPS. You will need
    an SSL certificate for the web server. A self-signed certificate may be
    even better than an official one, since your private key will be stored
    in the unencrypted partition, and this is not a public service anyway.
    If you don't know how to generate a self-signed certificate, have a look
    at the "SSL.txt" file. The remaining text assumes that you have stored
    your certificate as "boot.crt" and the private key as "boot.key".

 B. You may want to have a separate subdomain for the password page, so your
    ordinary web visitors don't find it. It will only be up for the minute
    it takes you to enter the password, but it's not intended for high load.
    Allocate a new Elastic IP and update your DNS to get the new subdomain.
    For example, you may want to add boot.example.com to the DNS records.

 C. The instructions below will create a system based on Ubuntu 16.04 LTS.
    Add "--system artful" to the command in step 6 if you want Artful
    instead of Xenial. The "--big-boot" option enables an alternative way of
    booting the instance, which installs a full Ubuntu system on /dev/xvda1.
    In this case, no changes are made to the system stored on the encrypted
    /dev/xvda2 partition. In theory, this will allow you to install any type
    of Linux distro there, instead of being limited to Ubuntu (or Debian).
    The downside is that /dev/xvda1 must be updated separately in that case.

-------------------------------------------------------------------------------

 1. Launch a new Ubuntu instance, so you don't delete anything important.
    This is just a build server, but the encrypted instance will have the
    same SSH key pair, instance type, availability zone, and architecture
    as this server, so you want to select them now.

 2. Connect to your Ubuntu instance and install cryptsetup:
      sudo apt-get -y install cryptsetup

 3. Upload all necessary files to a directory on the Ubuntu instance:
      activate.cgi                - password-fetching CGI script
      awsapi                      - used by the encroot script
      boot.crt                    - SSL certificate for the boot partition
      boot.key                    - private SSL key for the boot partition
      cryptsetup                  - initramfs hook script
      cryptsetup.sh               - initramfs cryptsetup replacement
      encroot                     - handles all necessary calls to the EC2 API
      hiding.gif                  - animated GIF used to hide text
      index.html                  - page that redirects to activate.cgi
      init.sh                     - script replacing /sbin/init in a Big Boot
      make_bozo_dir.sh            - bozohttpd home setup script
      make_encrypted_distro.sh    - does all the work on an attached volume
      pre_init.sh                 - script that gets the password in a Big Boot
      uecimage.gpg                - public key for Ubuntu images

 4. To avoid a lot of unnecessary manual work, put your AWS "Secret Access Key"
    and "Access Key ID" in the .awsapirc file as explained by "awsapi --man":
      touch ~/.awsapirc; chmod 600 ~/.awsapirc; nano ~/.awsapirc
      secretAccessKey: <your-secret-access-key>
      accessKeyId: <your-access-key-id>

 5. Create a security group for the encrypted instance. This group must allow
    HTTPS access, since this is how you will enter the password.

 6. Now, something like this should create and launch the instance for you:
    ./encroot --group "Security Group" boot.example.com

 7. The script will ask you for two passwords. Just follow the instructions.
    Expect a total running time of around 10 minutes on a micro instance.

 8. Visit https://boot.example.com/ in your web browser to unlock your root.
    The script automatically associates any Elastic IP address that this domain
    points to, but you must do this yourself the next time you boot the server.
    Once you have entered the decryption password, the system should boot.

 9. Don't forget to terminate your build instance when you're finished.

-------------------------------------------------------------------------------


ChangeLog.txt
2018-01-21:
 o Increased the length of dotted lines to handle longer resource IDs
 o Updated PV-GRUB kernels from version 1.04 to 1.05 in all supported regions
 o Added PV-GRUB kernels for Seoul, Mumbai, Canada, London, and Ohio
 o Added name filter to speed up selection of platform AMI during test
 o Added Seoul, Mumbai, Canada, London, and Ohio to the "regions" suite
 o Started building bozohttpd from source code since the package was removed
 o Updated partition selection to handle Debian AMIs with files in partition 2
 o Updated to use conf-hook for configuration of new cryptsetup packages
 o Fixed previously incorrect test index when individual tests are skipped
 o Removed code introduced as special cases for obsolete distros
 o Updated the EC2_API_VERSION to 2016-11-15 in encroot and test_encroot.sh
 o Removed the obsolete "newtype" and "oldtype" test suites
 o Changed the default version of Debian from 8 (Jessie) to 9 (Stretch)
 o Changed the default version of Ubuntu from 14.04 (Trusty) to 16.04 (Xenial)
 o Added "--bootsize" option to set an explicit size for the boot partition
 o Doubled default size of the boot partition to 2 GiB for "Big Boot" systems
 o Added support for systemd as a replacement for init in "Big Boot" systems
 o Fixed "Big Boot" when the primary network interface has not been defined
 o Added code to reinstall ifupdown package since Ubuntu switched to netplan
 o Changed EC2_ENDPOINT for Chinese regions to end in ".com.cn"; Thanks, Paul!
 o Corrected detection of fancy terminals by handling patterns like "xterm*"
 o Fixed moving of the mounted boot partition with "mount --move" on Xenial
 o Updated the recommended build platform to Ubuntu 16.04 LTS (Xenial Xerus)
 o Made temporary instances launch in the target subnet since some need a VPC
 o Added delay after launching instances to avoid InvalidInstanceID.NotFound
 o Added delay after registering images to avoid InvalidAMIID.NotFound
 o Fixed a bug that caused --version to always return the current date
 o Skipped tests of 32-bit Artful on t2.micro due to a Linux boot problem
 o Verified for Trusty, Xenial, Artful, Wheezy, Jessie, and Stretch
 o Precise, Utopic, Vivid, and Squeeze have lost upstream support

2015-05-28:
 o Fixed lookup of subnets + groups when "vpc-ff068f90" is given as "ff068f90"
 o Added support for hardware-based virtualization and made it the default
 o Added "--paravirtual" option to keep the original virtualization type
 o Started using the latest hd0 kernels instead of old hd00 AKI versions
 o Added PV-GRUB kernel IDs for the eu-central-1 (Frankfurt) region
 o Implemented signature version 4, needed in Frankfurt; Thank you, Chris!
 o Updated the EC2_API_VERSION to 2015-03-01 in encroot and test_encroot.sh
 o Added support for testing many different instance types in a test suite
 o Made it possible to use ENCROOT_HOSTS instead of DNS for test addresses
 o Starting testing Encroot with VPC instances in a separate test subnet
 o Changed the default type of test instances from t1.micro to t2.micro
 o Changed the default version of Debian from 7 (Wheezy) to 8 (Jessie)
 o Added the AWSAPI_USER_AGENT environment variable to give some flexibility
 o Changed "User-Agent" headers for Encroot to actually mention "Encroot"
 o Made it possible to choose an alternative port for the password page
 o Added "--ebs-type" and "--iops" options to simplify use of SSD volumes
 o Made the "debian-" prefix optional, so "jessie" works like "debian-jessie"
 o Updated awsapi to try again on temporary "Please try again shortly" errors
 o Made the "debian" and "ubuntu" forms work correctly with "/i386" suffixes
 o Tried to make unpacking and copying more robust by retrying on failure
 o Updated and extended the release suite from 56 to 168 tested combinations
 o Verified for Precise, Trusty, Utopic, Vivid, Squeeze, Wheezy, and Jessie
 o Lucid, Quantal, Raring, and Saucy have lost upstream support

2014-06-07:
 o Added "--dedicated" option to run VPC instances on single-tenant hardware
 o Fixed incorrect zone patterns for sa-east-1 and us-gov-west-1 kernel IDs
 o Updated PV-GRUB kernel IDs from version 1.03 to 1.04 in all nine regions
 o Made the changes needed for Ubuntu 14.04 LTS (Trusty Tahr)
 o Updated the default target to Trusty Tahr (20140528)
 o Added code to decrypt boot.key if the user forgets that step
 o Allowed "system" to be specified as a local file hierarchy
 o Allowed "system" to be an AMI, an instance, a snapshot, or an EBS volume
 o Added support for Debian as an alternative to Ubuntu
 o Added "--reg-params" option for extra parameters to the RegisterImage call
 o Added "--run-params" option for extra parameters to the RunInstances call
 o Updated the EC2_API_VERSION to 2014-02-01 in start_encrypted_instance.sh
 o Added "--fix-hook" option for more flexible handling of various distros
 o Allowed "system" to be "debian" or "ubuntu" with reasonable defaults
 o Made the separators fancier; use --no-lines to get the old dashed ones
 o Added bold (bright) and thin (dimmed) fonts; use --no-fonts to disable
 o Added a little color to the output; use --no-color to get rid of it
 o Added "--no-style" option to simplify "--no-color --no-fonts --no-lines"
 o Updated the README.txt file to reflect current reality and new features
 o Renamed "start_encrypted_instance.sh" to "encroot" for simplicity
 o Renamed "make_encrypted_ubuntu.sh" to "make_encrypted_distro.sh"
 o Made it possible to install Encroot with "./configure && make install"
 o Added "--version" option to easily identify the installed Encroot version
 o Added "--type" option for instance types different from the build instance
 o Added "--key" option for key pairs other than that of the build instance
 o Verified for Precise, Quantal, Raring, Saucy, Trusty, Squeeze, and Wheezy
 o Oneiric has lost upstream support; Lucid depended on bozohttpd from Natty

2013-06-20:
 o Added a PATH check to avoid hiding /sbin/cryptsetup with our own script

test_suites.sh
#!/bin/sh
###############################################################################
# This file contains alternative suites run by the test_encroot.sh script.    #
#                                                                             #
# It's included as is at the beginning of test_encroot.sh, but is a separate  #
# file to encourage users to add simple modifications and extra suites. This  #
# code is trivial, so I consider it to be in the public domain.               #
#                                                                             #
###############################################################################

SUITES="32_bits bigboot default distros failure
        regions release trivial"

SUITE_32_bits() {
    DESCRIPTION="32-bit test for each supported distro"
    REGIONS="us-east-1"
    PLATFORMS="xenial"
    TARGETS="trusty xenial artful wheezy jessie stretch"
    ARCHS="i386"
    TYPES="t1.micro t2.micro"
    OPTIONS="none"
}

SUITE_bigboot() {
    DESCRIPTION="Test with --big-boot on all distros"
    REGIONS="us-east-1"
    PLATFORMS="xenial"
    TARGETS="trusty xenial artful wheezy jessie stretch"
    ARCHS="x86_64"
    TYPES="t1.micro t2.micro"
    OPTIONS="--big-boot"
}

SUITE_default() {
    DESCRIPTION="64-bit Ubuntu + Debian (normal/big-boot)"
    REGIONS="us-east-1"
    PLATFORMS="xenial"
    TARGETS="ubuntu debian"
    ARCHS="x86_64"
    TYPES="t2.micro"
    OPTIONS="none --big-boot"
}

SUITE_distros() {
    DESCRIPTION="64-bit test for each supported distro"
    REGIONS="us-east-1"
    PLATFORMS="xenial"
    TARGETS="trusty xenial artful wheezy jessie stretch"
    ARCHS="x86_64"
    TYPES="t1.micro t2.micro"
    OPTIONS="none"
}

SUITE_failure() {
    DESCRIPTION="A failure to care about"
    REGIONS="us-east-1"
    PLATFORMS="xenial"
    TARGETS="jessie"
    ARCHS="i386"
    TYPES="t2.micro"
    OPTIONS="none"
}

SUITE_regions() {
    DESCRIPTION="64-bit test for each available region"
    REGIONS="ap-northeast-1  ap-northeast-2  ap-south-1
             ap-southeast-1  ap-southeast-2  ca-central-1
             eu-central-1    eu-west-1       eu-west-2
             sa-east-1       us-east-1       us-east-2
             us-west-1       us-west-2"
    PLATFORMS="xenial"
    TARGETS="xenial"
    ARCHS="x86_64"
    TYPES="t2.micro"
    OPTIONS="none"
}

SUITE_release() {
    DESCRIPTION="The official Encroot release suite"
    REGIONS="us-east-1 ap-northeast-1 eu-central-1"
    PLATFORMS="xenial"
    TARGETS="trusty xenial artful wheezy jessie stretch"
    ARCHS="i386 x86_64"
    TYPES="t1.micro t2.micro"
    OPTIONS="none --big-boot"
}

SUITE_trivial() {
    DESCRIPTION="Tests only 64-bit Xenial Xerus"
    REGIONS="us-east-1"
    PLATFORMS="xenial"
    TARGETS="xenial"
    ARCHS="x86_64"
    TYPES="t2.micro"
    OPTIONS="none"
}

###############################################################################


pre_init.sh
#!/bin/sh
###############################################################################
# This script is used to provide the password for an encrypted file system by #
# launching a minimal web server from the init script. It should be stored as #
# /etc/ec2/pre_init.sh on the pivoting AMI image to run in the main script.   #
#                                                                             #
# See this thread for some background information:                            #
#   http://developer.amazonwebservices.com/connect/thread.jspa?threadID=24091 #
#                                                                             #
###############################################################################
#                                                                             #
# Copyright (c) 2009, 2011, 2014-2015 Henrik Gulbrandsen <henrik@gulbra.net>  #
#                                                                             #
# This software is provided 'as-is', without any express or implied warranty. #
# In no event will the authors be held liable for any damages arising from    #
# the use of this software.                                                   #
#                                                                             #
# Permission is granted to anyone to use this software for any purpose,       #
# including commercial applications, and to alter it and redistribute it      #
# freely, subject to the following restrictions:                              #
#                                                                             #
# 1. The origin of this software must not be misrepresented; you must not     #
#    claim that you wrote the original software. If you use this software     #
#    in a product, an acknowledgment in the product documentation would be    #
#    appreciated but is not required.                                         #
#                                                                             #
# 2. Altered source versions must be plainly marked as such, and must not be  #
#    misrepresented as being the original software.                           #
#                                                                             #
# 3. This notice may not be removed or altered from any source distribution,  #
#    except that more "Copyright (c)" lines may be added to already existing  #
#    "Copyright (c)" lines if you have modified the software and wish to make #
#    your changes available under the same license as the original software.  #
#                                                                             #
###############################################################################

### Variables reserved for this script ########################################

pi_priv="/etc/ssl/private/boot.key"
pi_cert="/etc/ssl/certs/boot.crt"
pi_host="boot.example.com"
pi_home="/var/bozo"
pi_port="443"

pi_fifo="$pi_home/tmp/fifo"
pi_open=false
pi_stop=false
pi_udev=false

### Functions reserved for this script ########################################

pi_startWebServer() {
    echo "Removing stale GIFs and locks"
    rm -f "$pi_home"/data/hiding[0-9]*.gif
    rm -f "$pi_home/tmp/lock"

    if ! ps -e | grep udev | grep -qv grep; then
        echo "Starting udev"
        udevd --daemon
        pi_udev=true
    fi

    echo "Starting web server"
    bozohttpd -b -s -S "Introdus 1.0" -U bozo -t $pi_home -n -c /cgi-bin \
        -I $pi_port -Z $pi_cert $pi_priv data $pi_host

    # Remember the start
    pi_open=true
}

pi_stopWebServer() {
    if $pi_udev; then
        echo "Stopping udev"
        pkill udevd
    fi

    echo "Stopping web server"
    pkill bozohttpd
    pi_open=false
}

### API called from the init script ###########################################

fetch_password() {
    # Prepare the CGI script...
    if ! $pi_open; then
        start_net; pi_startWebServer;
    fi

    # ...and wait for its data
    echo "Waiting for password"
    password=$(cat $pi_fifo)
    echo "Got password"
}

accept_password() {
    echo "Accepting password"
    echo "OK" > $pi_fifo;
    sleep 1; pi_stopWebServer;
}

reject_password() {
    echo "Rejecting password"
    echo "BAD" > $pi_fifo;
}

###############################################################################


cryptsetup.sh
#!/bin/sh
###############################################################################
# This script is used to provide the password for an encrypted file system by #
# launching a minimal web server. It is meant to replace /sbin/cryptsetup for #
# an initramfs. The original cryptsetup should be moved to /sbin/cs instead.  #
#                                                                             #
###############################################################################
#                                                                             #
# Copyright (c) 2011, 2014-2015 Henrik Gulbrandsen <henrik@gulbra.net>        #
#                                                                             #
# This software is provided 'as-is', without any express or implied warranty. #
# In no event will the authors be held liable for any damages arising from    #
# the use of this software.                                                   #
#                                                                             #
# Permission is granted to anyone to use this software for any purpose,       #
# including commercial applications, and to alter it and redistribute it      #
# freely, subject to the following restrictions:                              #
#                                                                             #
# 1. The origin of this software must not be misrepresented; you must not     #
#    claim that you wrote the original software. If you use this software     #
#    in a product, an acknowledgment in the product documentation would be    #
#    appreciated but is not required.                                         #
#                                                                             #
# 2. Altered source versions must be plainly marked as such, and must not be  #
#    misrepresented as being the original software.                           #
#                                                                             #
# 3. This notice may not be removed or altered from any source distribution,  #
#    except that more "Copyright (c)" lines may be added to already existing  #
#    "Copyright (c)" lines if you have modified the software and wish to make #
#    your changes available under the same license as the original software.  #
#                                                                             #
###############################################################################

. /scripts/functions

### Variables reserved for this script ########################################

cs_priv="/etc/ssl/private/boot.key"
cs_cert="/etc/ssl/certs/boot.crt"
cs_host="boot.example.com"
cs_home="/bozo"
cs_port="443"

cs_fifo="$cs_home/tmp/fifo"
cs_open=false

### Web Server Handling #######################################################

cs_startWebServer() {
    echo "Removing stale GIFs and locks"
    rm -f "$cs_home"/data/hiding[0-9]*.gif
    rm -f "$cs_home/tmp/lock"

    echo "Starting web server"
    bozohttpd -b -s -S "Introdus 1.0" -U www-data -t $cs_home -n -c /cgi-bin \
        -I $cs_port -Z $cs_cert $cs_priv data $cs_host

    # Remember the start
    cs_open=true
}

cs_stopWebServer() {
    echo "Stopping web server"
    pkill bozohttpd
    cs_open=false
}

### Password Handling #########################################################

fetch_password() {
    # Prepare the CGI script...
    if ! $cs_open; then
        modprobe af_packet
        IPOPTS=dhcp; DEVICE=eth0
        configure_networking;
        cs_startWebServer;
    fi

    # ...and wait for its data
    echo "Waiting for password"
    password=$(cat $cs_fifo)
    echo "Got password"
}

accept_password() {
    echo "Accepting password"
    echo "OK" > $cs_fifo;
    sleep 1; cs_stopWebServer;
}

reject_password() {
    echo "Rejecting password"
    echo "BAD" > $cs_fifo;
}

### Core Script ###############################################################

if [ "_$1" = _isLuks ]; then
    exec /sbin/cs "$@"
fi

echo "ROOT=/dev/mapper/cryptroot" >> /conf/param.conf
echo "Replaced cryptsetup"

while fetch_password; do
    if printf "%s" "$password" | /sbin/cs "$@"; then
        accept_password; break;
    else
        reject_password;
    fi
done

###############################################################################


init.sh
#!/bin/sh
###############################################################################
# This script is used to boot an EBS image from a minimal AMI in Amazon EC2.  #
# It should be installed as a replacement for /sbin/init on the boot AMI.     #
#                                                                             #
# More detailed instructions are found in a separate README.txt file.         #
#                                                                             #
# See this thread for some background information:                            #
#   http://developer.amazonwebservices.com/connect/thread.jspa?threadID=24091 #
#                                                                             #
###############################################################################
#                                                                             #
# Copyright (c) 2009, 2011, 2013 Henrik Gulbrandsen <henrik@gulbra.net>       #
#                                                                             #
# This software is provided 'as-is', without any express or implied warranty. #
# In no event will the authors be held liable for any damages arising from    #
# the use of this software.                                                   #
#                                                                             #
# Permission is granted to anyone to use this software for any purpose,       #
# including commercial applications, and to alter it and redistribute it      #
# freely, subject to the following restrictions:                              #
#                                                                             #
# 1. The origin of this software must not be misrepresented; you must not     #
#    claim that you wrote the original software. If you use this software     #
#    in a product, an acknowledgment in the product documentation would be    #
#    appreciated but is not required.                                         #
#                                                                             #
# 2. Altered source versions must be plainly marked as such, and must not be  #
#    misrepresented as being the original software.                           #
#                                                                             #
# 3. This notice may not be removed or altered from any source distribution,  #
#    except that more "Copyright (c)" lines may be added to already existing  #
#    "Copyright (c)" lines if you have modified the software and wish to make #
#    your changes available under the same license as the original software.  #
#                                                                             #
###############################################################################

### Important variables #######################################################

# num is 16 * drive_number + partition (sda is SCSI disk 0; sdf is disk 5)
dev="/dev/sda2"
num="2"

# This external script can be used to perform various extra tasks, such as
# automatically attaching EBS volumes or fetching passwords from the net.
pre_init="/etc/ec2/pre_init.sh"

# The passphrase for an encrypted volume is fetched from EC2 user-data, but if
# one-time passwords are used, it will no longer be valid after a warm reboot.
# This file can be used to carry the new passphrase across reboots. It will be
# deleted by shred(1) as soon as this script has read it, for what it's worth.
key_file="/var/ec2/key_file.txt"

# These make things easier
new_root="/vol" # Location of the EBS drive when mounted in the AMI drive
old_root="/mnt" # Location of the AMI drive when mounted in the EBS drive
cleanup=""      # Shell commands run after chroot; must end with a semicolon

### Help functions ############################################################

# Some state
networking=false

# Starts networking if necessary
start_net() {
    if ! $networking; then
        mkdir -p /var/run/network
        ifup -a
        networking=true
    fi
}

# Stops networking if necessary
stop_net() {
    if $networking; then
        ifdown -a
        networking=false;
    fi
}

# These may be implemented by the $pre_init script
fetch_password()  { return false; }
accept_password() { return false; }
reject_password() { return false; }

### Fix the bootstrap environment #############################################

# These are needed on some systems
PATH=/sbin:/bin:/usr/sbin:/usr/bin:$PATH
unset SHLVL

# Make writable just in case files are missing...
echo "Remounting writable."
mount -o remount,rw /
mkdir -p $new_root

# These are needed for decryption
modprobe dm-crypt
modprobe sha256

# Include the optional $pre_init script
if [ -e $pre_init ]; then
    . $pre_init
fi

### Convert num to a partition name ###########################################

letters="abcdefghijklmnop"
drive=$((num/16))

while [ $drive -gt 0 ]; do
    letters="${letters#?}"
    drive=$((drive-1))
done

letter=${letters%${letters#?}}
digits=$((num%16))
part="${letter}${digits#0}"

### Waiting for the root device ###############################################

# The udevd is not started at this point (unless $pre_init started it)
for devNode in "8:/dev/sd${part}" "202:/dev/xvd${part}"; do
    major="${devNode%%:*}"; devNode="${devNode#*:}"

    if [ -e $devNode ]; then
        echo "Found existing device node: $devNode"
    else
        echo "Creating device node: $devNode"
        mknod -m 660 $devNode b $major $num
        cleanup="${cleanup}rm $devNode;"
    fi
done

# Find the right device for a given LABEL or UUID
if [ "${dev#LABEL=}" != "$dev" ]; then
    dev=$(blkid -L "${dev#LABEL=}")
elif [ "${dev#UUID=}" != "$dev" ]; then
    dev=$(blkid -U "${dev#UUID=}")
fi

# Try SOMETHING if that failed
if [ -z "$dev" ]; then
    echo "No suitable device found; using fallback"
    dev="/dev/sd${part}"
fi

# This may be of interest if things go wrong
echo "Selected device node: $dev"

# It may take a while for the root device to show up
while ! blockdev --getsize $dev > /dev/null 2>&1; do
    echo "Waiting for the root device."
    sleep 10
done

echo "Detected root device."
cryptCount=0

### Optional decryption #######################################################

# Handle encrypted file systems
while cryptsetup isLuks $dev; do

    echo "Encrypted disk; looking for the password."
    cryptCount=$((cryptCount+1))

    # Try this function from $pre_init first
    if [ -z "$password" ]; then
        fetch_password;
    fi

    # The password may have been saved during a warm reboot
    if [ -z "$password" ] && [ -e $key_file.txt ]; then
        password=$(cat $key_file)
        shred -zu $key_file
    fi

    # Otherwise, it should be given as user data...
    if [ -z "$password" ]; then
        data_url="http://169.254.169.254/2009-04-04/user-data"
        curl="curl --retry 3 --silent --show-error --fail"
        start_net; password=$($curl $data_url | head -1 | sed 's/.*Key: //')
    fi

    echo "Starting decryption."
    name="luks$cryptCount"

    # This should always work if the password is correct
    if ! printf "$password" | cryptsetup --key-file=- luksOpen $dev $name; then

        echo "Decryption failed."

        # First see if $pre_init can handle it
        if reject_password; then
            cryptCount=$((cryptCount-1))
            password=""
            continue
        fi

        # otherwise, just hang forever...
        while true; do
            sleep 60
        done;
    fi

    echo "Decryption worked."

    # Use the decrypted file system
    dev="/dev/mapper/$name"
    if [ ! -e $dev ]; then
        mknod -m 660 $dev b 253 $((cryptCount-1))
    fi
done

# Cleanup from $pre_init
accept_password;

### Pivot to the new root #####################################################

# Tear down the network if necessary
stop_net;

# This will hopefully work...
echo "Mounting $dev as new root."
mount -o ro $dev $new_root

# Pivot to the new root
cd $new_root; pivot_root . ${old_root#/}

# Move mounted file systems to the new root
for dir in /dev /proc /run /sys; do
    mount --move ${old_root}${dir} ${dir} 2> /dev/null || true
done

# Get rid of $old_root; it can be mounted from /dev/sda1 later
cleanup="${cleanup}umount ${old_root}; "

# Start the real init with redirects so the old /dev can be released
exec chroot . /bin/sh -c "${cleanup}exec /sbin/init $*" \
    < /dev/console > /dev/console 2>&1

###############################################################################

